### ISSUE TITLE
how to avoid the precision loss(float32) caused by the gradient accumulation of Ring Allreduce in the case of ddp

### ISSUE THREAD

**Issue by lezasantaizi** [pytorch/pytorch#48576]
### how to avoid the precision loss(float32) caused by the gradient accumulation of Ring Allreduce in the case of ddp.

How to avoid the precision loss(float32) caused by the gradient accumulation of Ring Allreduce in the case of ddp

when run model in single gpu twice, the weight is always same; 
when run model in ddp twice , the weight is different in grad apply. 

I suspect that the gradient error is accumulated in the Ring Allreduce.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd

**Comment by osalpekar** [comment]
@lezasantaizi Are you using NCCL or Gloo as the communication backend for DDP? Also just to confirm, everything else is held constant across the runs (model weights initialization, ordering/shuffling on the data, etc.)?

I believe there is some non-determinism introduced by NCCL since it computes the best way of performing allreduce each time, and so the exact configuration such as the number of rings may be different for different runs (similar post here: https://discuss.pytorch.org/t/does-ddp-make-computations-non-deterministic/100641). There are ways of fixing the number of rings and making NCCL more deterministic using configurable environment variables, but it would be best to find out more from the NCCL folks directly.

cc @SciPioneer - Are you aware of any other potential numerical non-determinism when using NCCL + DDP?

**Comment by lezasantaizi** [comment]
@osalpekar ,Thanks for your reply!
My confusion are the same as [yanz0920](https://discuss.pytorch.org/t/does-ddp-make-computations-non-deterministic/100641)
,but how to fix the rings ? it looks like the questioner is solved the problem. 
![image](https://user-images.githubusercontent.com/7878465/100694061-2f4dc100-33c9-11eb-83ed-3396f2720d45.png)

**Comment by osalpekar** [comment]
I believe `NCCL_RINGS` environment variable allows you to explicitly configure the rings used by NCCL, but this is a better question to ask in the NCCL GitHub directly

**Comment by lezasantaizi** [comment]
@osalpekar ,Thanks! 
I will to ask the NCCL forks in the NCCL GitHub.

**Comment by pritamdamania87** [comment]
> when run model in ddp twice , the weight is different in grad apply.

@lezasantaizi How many GPUs are you using here and are you tuning the learning rate/batch size accordingly to accommodate for multiple GPUs?

Sharing this note from the [DDP documentation](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel): 

> When a model is trained on M nodes with batch=N, the gradient will be M times smaller when compared to the same model trained on a single node with batch=M*N (because the gradients between different nodes are averaged). You should take this into consideration when you want to obtain a mathematically equivalent training process compared to the local training counterpart.

**Comment by lezasantaizi** [comment]
> > when run model in ddp twice , the weight is different in grad apply.
> 
> @lezasantaizi How many GPUs are you using here and are you tuning the learning rate/batch size accordingly to accommodate for multiple GPUs?
> 
> Sharing this note from the [DDP documentation](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel):
> 
> > When a model is trained on M nodes with batch=N, the gradient will be M times smaller when compared to the same model trained on a single node with batch=M*N (because the gradients between different nodes are averaged). You should take this into consideration when you want to obtain a mathematically equivalent training process compared to the local training counterpart.

there are 4gpus in each ; and my confusion is the problem of deterministic model. @osalpekar has solved my confusion;
